server/src/pull/pull.ts

```diff
@@ -0,0 +1,258 @@
+import {z} from 'zod';
+import type {PatchOperation, PullResponse, PullResponseOKV1} from 'replicache';
+import {transact} from '../pg';
+import {getClientGroupForUpdate, putClientGroup, searchClients} from '../data';
+import type Express from 'express';
+import {hasNextPage, isPageEmpty, readNextPage} from './next-page';
+import {
+  CVR,
+  dropCVREntries,
+  getCVR,
+  putCVR,
+  recordDeletes,
+  recordUpdates,
+} from './cvr';
+import {PartialSyncState, PARTIAL_SYNC_STATE_KEY} from 'shared';
+
+const cookieSchema = z.object({
+  order: z.number(),
+  clientGroupID: z.string(),
+});
+
+export type Cookie = z.infer<typeof cookieSchema>;
+
+export const pullRequest = z.object({
+  clientGroupID: z.string(),
+  cookie: z.union([cookieSchema, z.null()]),
+});
+
+export async function pull(
+  requestBody: Express.Request,
+): Promise<PullResponse> {
+  const start = performance.now();
+  console.log(`Processing pull`, JSON.stringify(requestBody, null, ''));
+  const pull = pullRequest.parse(requestBody);
+  const {clientGroupID} = pull;
+
+  const {clientChanges, nextPage, prevCVR, nextCVR} = await transact(
+    async executor => {
+      // Get a write lock on the client group first to serialize with other
+      // requests from the CG and avoid deadlocks.
+      const baseClientGroupRecord = await getClientGroupForUpdate(
+        executor,
+        clientGroupID,
+      );
+      const prevCVR = pull.cookie
+        ? await getCVR(executor, pull.cookie.clientGroupID, pull.cookie.order)
+        : undefined;
+      const baseCVR: CVR = prevCVR ?? {
+        clientGroupID,
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
+      // Getting an old order from the client means that the client is possibly missing the data
+      // from future orders.
+      //
+      // The other possbility is that multiple tabs are syncing (why are tabs not coordinating who does sync?)
+      // Given that, dropping entries can cause to tabs could starve one another!! I've updated the frontend
+      // so pulls can only be issued from one tab (see `useExclusiveEffect`). The other solution here is to not
```

Comment:
We looked at web locks extensively and decided not to rely on them. The semantics around tab freezing, bfcache, and similar are super _super_ weird, and we have low confidence they actually work right and across browsers.

I think we should continue to allow multiple tabs to sync because it's robust, but we should prevent _background_ tabs from syncing.

I don't think the starvation thing you mention will happen.

Reply:
> but we should prevent background tabs from syncing.

Just via something like `document.hidden`?

Reply:
Yes, that's what I'm thinking.